lib/codegen/src/verifier/mod.rs

```diff
@@ -344,21 +344,91 @@ impl<'a> Verifier<'a> {
+                    }
                 }
+                ir::GlobalValueData::Deref { base, .. } => {
+                    if let Some(isa) = self.isa {
+                        let base_type = self.func.global_values[base].global_type(isa);
+                        let pointer_type = isa.pointer_type();
+                        if base_type != pointer_type {
+                            report!(
+                                errors,
+                                gv,
+                                "deref base {} has type {}, which is not the pointer type {}",
+                                base,
+                                base_type,
+                                pointer_type
+                            );
+                        }
+                    }
+                }
+                _ => {}
             }
         }
 
         // Invalid global values shouldn't stop us from verifying the rest of the function
         Ok(())
     }
 
+    fn verify_heaps(&self, errors: &mut VerifierErrors) -> VerifierStepResult<()> {
+        if let Some(isa) = self.isa {
+            for (heap, heap_data) in &self.func.heaps {
+                let base = heap_data.base;
+                if !self.func.global_values.is_valid(base) {
+                    return nonfatal!(errors, heap, "invalid base global value {}", base);
+                }
+
+                let pointer_type = isa.pointer_type();
+                let base_type = self.func.global_values[base].global_type(isa);
+                if base_type != pointer_type {
+                    report!(
+                        errors,
+                        heap,
+                        "heap base has type {}, which is not the pointer type {}",
+                        base_type,
+                        pointer_type
+                    );
+                }
+
+                match heap_data.style {
+                    ir::HeapStyle::Dynamic { bound_gv, .. } => {
+                        if !self.func.global_values.is_valid(bound_gv) {
+                            return nonfatal!(
+                                errors,
+                                heap,
+                                "invalid bound global value {}",
+                                bound_gv
+                            );
+                        }
+
+                        let index_type = heap_data.index_type;
+                        let bound_type = self.func.global_values[bound_gv].global_type(isa);
+                        if index_type != bound_type {
+                            report!(
+                                errors,
+                                heap,
+                                "heap index type {} differs from the type of its bound, {}",
+                                index_type,
+                                bound_type
+                            );
+                        }
+                    }
+                    _ => {}
```

Comment:
I wonder if the default arm is ok for this case. Worrying about that it might be missed if a new heap style were added.